ion/textutils.go

```diff
@@ -342,39 +342,97 @@ func parseTimestamp(val string) (time.Time, error) {
 	}
 
 	day, err := strconv.ParseInt(val[8:10], 10, 32)
-	if err != nil {
+	if err != nil || day < 1 {
 		return invalidTimestamp(val)
 	}
 
 	if len(val) == 10 || (len(val) == 11 && (val[10] == 't' || val[10] == 'T')) {
 		// yyyy-mm-dd or yyyy-mm-ddT
-		return time.Date(int(year), time.Month(month), int(day), 0, 0, 0, 0, time.UTC), nil
+		return tryCreateTimeDate(val, year, month, day, time.UTC)
 	}
 	if val[10] != 't' && val[10] != 'T' {
 		return invalidTimestamp(val)
 	}
 
+	// At this point timestamp must have hour:minute
 	if len(val) < 17 {
 		return invalidTimestamp(val)
 	}
-	if val[16] != ':' {
-		return time.Parse("2006-01-02T15:04Z07:00", val)
+	if val[16] == 'z' || val[16] == 'Z' {
+		return time.Parse("2006-01-02T15:04Z", val)
```

Comment:
As a Go newbie, this line really confused me. I had to read [the `Parse()` docs](https://golang.org/pkg/time/#Parse) to understand why we were writing out a seemingly arbitrary date from the winter of 2006.

> Parse parses a formatted string and returns the time value it represents. The layout defines the format by showing how the reference time, defined to be
> 
>     Mon Jan 2 15:04:05 -0700 MST 2006
> 
> would be interpreted if it were the value; it serves as an example of the input format. The same interpretation will then be made to the input string. 
>
>  Predefined layouts ANSIC, UnixDate, **RFC3339** and others describe standard and convenient representations of the reference time. For more information about the formats and the definition of the reference time, see the documentation for ANSIC and the other constants defined by this package. Also, the executable example for Time.Format demonstrates the working of the layout string in detail and is a good reference. 

I'd like to use `time.RFC3339` instead of the string literal `"2006-01-02T15:04Z"` for simplicity.

Reply:
If it turns out that `time.RFC3339` behaves differently from the string literal version, a comment explaining what the string literal means would also be fine.

Reply:
Yeah, it's a clever way of doing format strings and also _totally_ confusing if you're not aware of that convention. :/ `time.RFC3339` expects the string being parsed to include seconds, which is why this one was a literal. I'd be +1 for adding a comment explaining what it's doing.

Reply:
Added some explanation.
__  
We also can add something like:
```
const (
	Layout_yyyy_mm_ddThh_mmZhh_mm         = "2006-01-02T15:04Z07:00"   // Comment
	Layout_yyyy_mm_ddThh_mmZ              = "2006-01-02T15:04Z"             // Comment
	Layout_yyyy_mm_ddThh_mm_ss_nsecZhh_mm = time.RFC3339Nano // "2006-01-02T15:04:05.999999999Z07:00"
	Layout_yyyy_mm_ddThh_mm_ssZhh_mm      = time.RFC3339     // "2006-01-02T15:04:05Z07:00"
)
```

and then use these constants: ``` return time.Parse(Layout_yyyy_mm_ddThh_mmZhh_mm, val) ```

Let me know which one is preferred.

Reply:
Creating named constants for this is a good idea. :thumbsup: